PyEMD/significancetest.py

```diff
@@ -0,0 +1,127 @@
+
+
+#helper function: normalize data
+def normalize(data):
+    mean = sum(data)/len(data)
+    dmin = min(data)
+    dmax = max(data)
+    ret = np.array([])
+    for val in data:
+        ret = np.append(ret, (val-mean)/(dmax-dmin))
+    return ret
+
+#helper function: find zero-crossings in IMF, used for finding mean period of IMF
+def sign_change(data):
+    l = len(data)
+    counter=0
+    for x in range(l-1):
+        if (data[x]>0 or data[x] == 0) and data[x+1] < 0:
+            counter+=1
+    return counter
+
+#helper function: Find mean period of an IMF
+def mean_period(data):
+    s = sign_change(data)
+    if s != 0:
+        return len(data)/s
+    else:
+        return len(data)
+
+#helper function: find energy of signal/IMF
+def energy(data):
+    return sum(pow(data, 2))
+
+#helper function: find IMF significance in 'a priori' test
+def significance_apriori(energy_density, T, N, alpha):
+    k = abs(stats.norm.ppf((1-alpha)/2))
+    if (energy_density <= (-math.log(T) + k*(math.sqrt(2/N)*math.exp(math.log(T)/2)))) and (energy_density >= (-math.log(T) - k*(math.sqrt(2/N)*math.exp(math.log(T)/2)))):
+        return False
+    else:
+        return True
+
+#helper function: find significance in 'a posteriori' test
+def significance_aposteriori(scaled_energy_density, T, N, alpha):
+    k = abs(stats.norm.ppf((1-alpha)/2))
+    if (scaled_energy_density <= (-math.log(T) + k*(math.sqrt(2/N)*math.exp(math.log(T)/2)))):
+        return False
+    else:
+        return True
+
+def whitenoisecheck(data: np.ndarray, method: str='EMD',  test: str='aposteriori', rescaling_imf: int=1, alpha: float= 0.95):
+    '''
+    References
+    --------------
+    -- Zhaohua Wu, and Norden E. Huang. “A Study of the Characteristics of White Noise Using the Empirical Mode Decomposition Method.” Proceedings: Mathematical, Physical and Engineering Sciences, vol. 460, no. 2046, The Royal Society, 2004, pp. 1597–611, http://www.jstor.org/stable/4143111.
+
+    Examples
+    --------------
+    >>> import numpy as np
+    >>> from PyEMD.significancetest import whitenoisecheck
+    >>> T = np.linspace(0, 1, 100)
+    >>> S = np.sin(2*2*np.pi*T)
+    >>> significant_imfs = whitenoisecheck(S, test='apriori')
+    >>> significant_imfs
+    {1: 0, 2: 1}
+    >>> type(significant_imfs)
+    <class 'dict'>
+    '''
+    assert alpha > 0 and alpha < 1, "alpha value should be in between (0,1)"
+    assert method == 'EMD' or method == 'EEMD' or method == 'CEEMDAN', 'Invalid decomposition method'
+    assert test == 'apriori' or test == 'aposteriori', 'Invalid test type'
+    assert isinstance(data, np.ndarray), 'Invalid Data type, Pass a numpy.ndarray'
+
+    N = len(data)
+    if method == 'EMD':
+        emd = EMD()
+        IMFs = emd.emd(normalize(data))
+    elif method == 'EEMD':
+        eemd = EEMD()
+        IMFs = eemd.eemd(normalize(data))
+    elif method == 'CEEMDAN':
+        ceemdan = CEEMDAN()
+        IMFs = ceemdan.ceemdan(normalize(data))
```

Comment:
The benefit of this method is to perform an additional test after decomposing so it's rather assumed that people will have `IMFs`. Besides, decomposition classes here aren't configurable which limits the use.

Please assume that the method `whitenoise_check` receives `IMFs`.

Reply:
This may be a challenge.
The Significance test for IMFs has the following necessary steps:
1.  **Normalize the Original time-series**
2. Decompose the Normalized time-series
3. Use the IMFs from the decomposed normalized time-series for whitenoise_check.

If I am correct, the decomposition methods do not necessarily normalize the data at the start, thereby skipping the 1st step which is essentially required in `whitenoise_check`.
The issue can be somewhat resolved in `EMD` as by summing up the IMFs, we can get back the original time-series but the same is not true for EEMD and CEEMDAN as noise is added in the process of decomposition.
So, taking just the IMFs as input may not be feasible.

It is true that the way the function is implemented, it limits the use of the decomposition classes, thanks for pointing that out. One way to solve this would be to allow the parameters of decomposition classes as parameters of `whitenoise_check`. But I am still trying to consider if there is a better approach to get this done. Do let me know if you can think of any better approach or if I may implement this.

Reply:
I understand your concern about fulfilling certain conditions before using the method but there's a limit to what we can do. Everyone can take a random method and use it outside of its original intention, e.g. you can take a method that computes Student Test on uniformly distributed data. You shouldn't, because it doesn't make sense, but you can. Here is similar; we provide a useful tool but we have no control over how it'll eventually be used.

My take is that is should be enough to update documentation (docstring) and state that IMFs are supposed to be computed from normalized signal.

Reply:
Okay, I got your point. Have made the suggested changes.